### DISCUSSION TITLE
bootls integration

### DISCUSSION THREAD

**Discussion by jvalkeal** [nvim-java/nvim-java#99] [General]
I wanted to open this thread to keep boot-ls related discussions in one place.

**Comment by jvalkeal** [comment]
Dispatching messages between `jdtls` and `bootls`.

I'm having some trouble to understand how I could do thing without going full blocking way.

If I'm adding handler for `bootls` using something like:
```
['sts/addClasspathListener'] = on_sts_addClasspathListener
```
What I'd like to accomplish is to return response back to `bootls` only when `execute_command` returns from `jdtld`.
```
local JdtlsClient = require('java-core.ls.clients.jdtls-client')
local async = require('java-core.utils.async').sync

local function on_sts_addClasspathListener(err, result, ctx, config)
	local client = lsp.get_jdtls_client()
	log.debug('XXX client: ', client)
	log.debug('XXX result: ', result)
	local ret = async(function()
		local ret1 = JdtlsClient:new({client = client}):execute_command('sts.java.addClasspathListener', result.callbackCommandId)
		log.debug('XXX ret1: ', ret1)
		return ret1
	  end)
	  .run()
	log.debug('XXX ret: ', ret)
  return {}
end
```

I tried above but then realised that this async stuff may not be meant for this. Above is logging:
```
[DEBUG Sun 31 Dec 2023 15:30:19 GMT] ...java-core/lua/lspconfig/server_configurations/bootls.lua:59: XXX ret:  nil
[DEBUG Sun 31 Dec 2023 15:30:19 GMT] ...java-core/lua/lspconfig/server_configurations/bootls.lua:52: XXX ret1:  ok
```
Is it somewhat possible to get `ret1` value out from an async block before `on_sts_addClasspathListener` returns?

**Reply by s1n7ax** [reply]
@jvalkeal Unless the entire  `['sts/addClasspathListener'] = on_sts_addClasspathListener` statement is wrapped in `async` OR the one who's calling it wraps it, it's not possible. In sts4, it's similar. It returns a promise, so what ever calls that should be a async function.

```ts
    client.onRequest(addRequest, async (params: ClasspathListenerParams) => {
        return <ClasspathListenerResponse> await VSCode.commands.executeCommand("java.execute.workspaceCommand", "sts.java.addClasspathListener", params.callbackCommandId);
    });
```

Try using `request_sync` function.
>       • request_sync(method, params, timeout_ms, bufnr) Sends a request to the
>         server and synchronously waits for the response. This is a wrapper
>         around {client.request} Returns: { err=err, result=result }, a
>         dictionary, where `err` and `result` come from the |lsp-handler|. On
>         timeout, cancel or error, returns `(nil, err)` where `err` is a string
>         describing the failure reason. If the request was unsuccessful returns
>         `nil`.

**Reply by jvalkeal** [reply]
Thx, I came to same conclusion when I was reading that long async thread in neovim repo itself but wanted to ask to be sure.

**Comment by s1n7ax** [comment]
This is not implemented
